File: 2002.maximum-product-of-the-length-of-two-palindromic-subsequences.py
#
# @lc app=leetcode id=2002 lang=python3
#
# [2002] Maximum Product of the Length of Two Palindromic Subsequences
#

# @lc code=start
class Solution:
    '''
    2 <= s.length <= 12
    so we can use bitmask to represent which bits are used
    First thing to do: for each mask, determine if the mask represent a palindrome O(2 ^ 12)
    Then among the mask, determine if the subMasks of 0 ^ mask is inside the set
    '''
    '''
    Python program to decide is a string is a palindrome
    '''
    def isPalindrome(self, s):
        l, r = 0, len(s) - 1
        while l < r:
            if s[l] != s[r]: return False
            else: 
                l += 1
                r -= 1
        return True
    
    '''
    Method to return all subset of set bits of num1 while smaller than num2
    '''
    def setBits(self, num1, num2):
        result = set()
        temp = num1
        while temp > 0:
            if temp <= num2:
                result.add(temp)
            temp -= 1
            temp &= num1
        return result
    
    def maxProduct(self, s: str) -> int:
        n = len(s)
        palindromes =  {}
        for mask in range(1, 1 << n):
            sMask, count = "", 0
            for i in range(n - 1, -1, -1):
                if mask & (1 << i) > 0:
                    sMask += s[i]
                    count += 1
            if self.isPalindrome(sMask): palindromes[mask] = count
        
        result = 0
        for mask in range(1 << n):
            if mask in palindromes:
                antiMaskSet = self.setBits(((1 << n) - 1) ^ mask, 1 << n)
                for aMask in antiMaskSet:
                    if aMask in palindromes:
                        result = max(result, palindromes[mask] * palindromes[aMask])
        return result
        
# @lc code=end

File: 2002.maximum-product-of-the-length-of-two-palindromic-subsequences.cpp
/*
 * @lc app=leetcode id=2002 lang=cpp
 *
 * [2002] Maximum Product of the Length of Two Palindromic Subsequences
 */

// @lc code=start
#include <unordered_map>
#include <unordered_set>
#include <vector>
#include <string>
#include <algorithm>

using std::max;
using std::string;
using std::unordered_map;
using std::unordered_set;

class Solution {
public:
    /*
    2 <= s.length <= 12
    so we can use bitmask to represent which bits are used
    First thing to do: for each mask, determine if the mask represents a palindrome O(2 ^ 12)
    Then among the mask, determine if the subMasks of 0 ^ mask are inside the set
    */
    
    /*
    C++ program to decide if a string is a palindrome
    */
    bool isPalindrome(const string& s) {
        int l = 0, r = s.length() - 1;
        while (l < r) {
            if (s[l] != s[r]) {
                return false;
            } else {
                l++;
                r--;
            }
        }
        return true;
    }

    /*
    Method to return all subsets of set bits of num1 while smaller than num2
    */
    unordered_set<int> setBits(int num1, int num2) {
        unordered_set<int> result;
        int temp = num1;
        while (temp > 0) {
            if (temp <= num2) {
                result.insert(temp);
            }
            temp--;
            temp &= num1;
        }
        return result;
    }

    int maxProduct(string s) {
        int n = s.length();
        unordered_map<int, int> palindromes;
        for (int mask = 1; mask < (1 << n); mask++) {
            string sMask = "";
            int count = 0;
            for (int i = n - 1; i >= 0; i--) {
                if ((mask & (1 << i)) > 0) {
                    sMask += s[i];
                    count++;
                }
            }
            if (isPalindrome(sMask)) {
                palindromes[mask] = count;
            }
        }

        int result = 0;
        for (int mask = 1; mask < (1 << n); mask++) {
            if (palindromes.find(mask) != palindromes.end()) {
                unordered_set<int> antiMaskSet = setBits(((1 << n) - 1) ^ mask, 1 << n);
                for (int aMask : antiMaskSet) {
                    if (palindromes.find(aMask) != palindromes.end()) {
                        result = max(result, palindromes[mask] * palindromes[aMask]);
                    }
                }
            }
        }

        return result;
    }
};

// @lc code=end

